File: src/mcp_server/domain/rag_benchmarks.py
"""Deterministic open-source RAG quality metrics for local validation and UI tests."""

from __future__ import annotations

import math
from dataclasses import dataclass, replace
from typing import Literal

from mcp_server.domain.schemas import ChunkHit

ScoreKind = Literal["cosine", "rrf", "reranker"]
RetrievalMode = Literal["vector", "hybrid"]

# Minimum cosine similarity (query embedding vs passage embedding) to count a chunk as
# semantically relevant to the gold answer when computing precision@k.
DEFAULT_GOLD_SEMANTIC_PRECISION_THRESHOLD = 0.55


@dataclass(frozen=True, slots=True)
class ScoreThresholds:
    """Good/warn cutoffs for chunk scores, calibrated per score kind."""

    good: float
    warn: float


@dataclass(frozen=True, slots=True)
class RagEvaluationContext:
    """Run configuration affecting how retrieval metrics should be interpreted."""

    retrieval_mode: RetrievalMode
    retrieve_limit: int
    rerank_enabled: bool
    rerank_top_n: int
    effective_k: int
    score_kind: ScoreKind
    chunk_size: int | None = None
    chunk_overlap: int | None = None
    indexed_chunk_count: int | None = None
    hybrid_fts_active: bool = False
    rerank_applied: bool = False

    def as_dict(self) -> dict[str, str | int | bool | None]:
        return {
            "retrieval_mode": self.retrieval_mode,
            "retrieve_limit": self.retrieve_limit,
            "rerank_enabled": self.rerank_enabled,
            "rerank_top_n": self.rerank_top_n,
            "effective_k": self.effective_k,
            "score_kind": self.score_kind,
            "chunk_size": self.chunk_size,
            "chunk_overlap": self.chunk_overlap,
            "indexed_chunk_count": self.indexed_chunk_count,
            "hybrid_fts_active": self.hybrid_fts_active,
            "rerank_applied": self.rerank_applied,
        }


@dataclass(frozen=True, slots=True)
class RagBenchmarkScores:
    """Phrase-grounded retrieval quality metrics (no LLM judge required)."""

    phrase_coverage: float
    phrase_chunk_rate: float
    any_phrase_hit: float
    first_phrase_rank_reciprocal: float
    expected_phrase_count: int
    matched_phrase_count: int
    retrieved_chunk_count: int
    gold_semantic_relevance: float = 0.0
    mean_gold_semantic_relevance: float = 0.0
    gold_semantic_precision: float = 0.0
    gold_semantic_rank_reciprocal: float = 0.0

    def as_dict(self) -> dict[str, float | int]:
        return {
            "phrase_coverage": self.phrase_coverage,
            "phrase_chunk_rate": self.phrase_chunk_rate,
            "any_phrase_hit": self.any_phrase_hit,
            "first_phrase_rank_reciprocal": self.first_phrase_rank_reciprocal,
            "expected_phrase_count": self.expected_phrase_count,
            "matched_phrase_count": self.matched_phrase_count,
            "retrieved_chunk_count": self.retrieved_chunk_count,
            "gold_semantic_relevance": self.gold_semantic_relevance,
            "mean_gold_semantic_relevance": self.mean_gold_semantic_relevance,
            "gold_semantic_precision": self.gold_semantic_precision,
            "gold_semantic_rank_reciprocal": self.gold_semantic_rank_reciprocal,
        }

    def with_semantic_gold_scores(
        self,
        *,
        gold_semantic_relevance: float,
        mean_gold_semantic_relevance: float,
        gold_semantic_precision: float,
        gold_semantic_rank_reciprocal: float,
    ) -> RagBenchmarkScores:
        """Return a copy with semantic gold-answer metrics attached."""
        return replace(
            self,
            gold_semantic_relevance=round(gold_semantic_relevance, 4),
            mean_gold_semantic_relevance=round(mean_gold_semantic_relevance, 4),
            gold_semantic_precision=round(gold_semantic_precision, 4),
            gold_semantic_rank_reciprocal=round(gold_semantic_rank_reciprocal, 4),
        )


@dataclass(frozen=True, slots=True)
class RetrievalProxyMetrics:
    """Retrieval-only proxies when no ground-truth phrases are available."""

    chunk_count: int
    mean_chunk_score: float
    max_chunk_score: float
    context_length_chars: int
    score_kind: ScoreKind
    effective_k: int

    def as_dict(self) -> dict[str, float | int | str]:
        return {
            "chunk_count": self.chunk_count,
            "mean_chunk_score": self.mean_chunk_score,
            "max_chunk_score": self.max_chunk_score,
            "context_length_chars": self.context_length_chars,
            "score_kind": self.score_kind,
            "effective_k": self.effective_k,
        }


def score_thresholds_for_kind(score_kind: ScoreKind) -> ScoreThresholds:
    """Return good/warn thresholds calibrated for the given chunk score semantics."""
    if score_kind == "rrf":
        return ScoreThresholds(good=0.02, warn=0.01)
    if score_kind == "reranker":
        # Cross-encoder logits are mapped through sigmoid before display.
        return ScoreThresholds(good=0.5, warn=0.02)
    return ScoreThresholds(good=0.75, warn=0.45)


def resolve_score_kind(
    *,
    rerank_applied: bool,
    retrieval_mode: RetrievalMode,
    hybrid_fts_active: bool,
) -> ScoreKind:
    """Infer chunk score semantics from the retrieval path taken."""
    if rerank_applied:
        return "reranker"
    if retrieval_mode == "hybrid" and hybrid_fts_active:
        return "rrf"
    return "cosine"


def cosine_similarity(left: list[float], right: list[float]) -> float:
    """Cosine similarity between two L2-normalized or arbitrary vectors."""
    if not left or not right or len(left) != len(right):
        return 0.0
    dot = sum(a * b for a, b in zip(left, right, strict=True))
    left_norm = math.sqrt(sum(a * a for a in left))
    right_norm = math.sqrt(sum(b * b for b in right))
    if left_norm == 0.0 or right_norm == 0.0:
        return 0.0
    return dot / (left_norm * right_norm)


def compute_semantic_gold_benchmarks(
    *,
    gold_embedding: list[float],
    chunk_embeddings: list[list[float]],
    relevance_threshold: float = DEFAULT_GOLD_SEMANTIC_PRECISION_THRESHOLD,
) -> tuple[float, float, float, float]:
    """Compute semantic relevance and precision vs a gold-answer embedding.

    Returns:
        (max_relevance, mean_relevance, precision_at_k, rank_reciprocal)
    """
    if not gold_embedding or not chunk_embeddings:
        return 0.0, 0.0, 0.0, 0.0

    similarities = [cosine_similarity(gold_embedding, vector) for vector in chunk_embeddings]
    max_relevance = max(similarities)
    mean_relevance = sum(similarities) / len(similarities)
    relevant_count = sum(1 for value in similarities if value >= relevance_threshold)
    precision = relevant_count / len(similarities)

    best_rank: int | None = None
    for index, value in enumerate(similarities, start=1):
        if value >= relevance_threshold:
            best_rank = index
            break
    rank_reciprocal = 1.0 / best_rank if best_rank is not None else 0.0

    return max_relevance, mean_relevance, precision, rank_reciprocal


def _normalize_phrases(phrases: list[str]) -> list[str]:
    return [phrase.strip().lower() for phrase in phrases if phrase.strip()]


def _chunk_contains_phrase(chunk: ChunkHit, phrase: str) -> bool:
    return phrase in chunk.content.lower()


def build_retrieval_haystack(chunks: list[ChunkHit], *, merged_context: str = "") -> str:
    """Lowercase chunk bodies for phrase matching (chunks are the source of truth)."""
    del merged_context  # retained for call-site compatibility; not used in matching
    return "\n".join(chunk.content.lower() for chunk in chunks)


def partition_phrase_matches(
    expected_phrases: list[str],
    *,
    chunks: list[ChunkHit],
    merged_context: str = "",
) -> tuple[list[str], list[str]]:
    """Return (matched, missing) original phrases based on chunk content only."""
    haystack = build_retrieval_haystack(chunks, merged_context=merged_context)
    matched: list[str] = []
    missing: list[str] = []
    for phrase in expected_phrases:
        if not phrase.strip():
            continue
        if phrase.lower() in haystack:
            matched.append(phrase)
        else:
            missing.append(phrase)
    return matched, missing


def missing_expected_phrases(
    expected_phrases: list[str],
    *,
    chunks: list[ChunkHit],
    merged_context: str = "",
) -> list[str]:
    """Return original phrases not found in retrieved chunks."""
    _, missing = partition_phrase_matches(
        expected_phrases,
        chunks=chunks,
        merged_context=merged_context,
    )
    return missing


def compute_rag_benchmarks(
    *,
    expected_phrases: list[str],
    chunks: list[ChunkHit],
    merged_context: str = "",
) -> RagBenchmarkScores:
    """Compute deterministic RAG benchmarks from expected phrases and retrieved chunks."""
    phrases = _normalize_phrases(expected_phrases)
    haystack = build_retrieval_haystack(chunks, merged_context=merged_context)

    if not phrases:
        return RagBenchmarkScores(
            phrase_coverage=1.0,
            phrase_chunk_rate=1.0 if chunks else 0.0,
            any_phrase_hit=1.0 if chunks else 0.0,
            first_phrase_rank_reciprocal=0.0,
            expected_phrase_count=0,
            matched_phrase_count=0,
            retrieved_chunk_count=len(chunks),
        )

    matched = [phrase for phrase in phrases if phrase in haystack]
    phrase_coverage = len(matched) / len(phrases)

    relevant_chunks = 0
    first_relevant_rank: int | None = None
    for index, chunk in enumerate(chunks, start=1):
        if any(_chunk_contains_phrase(chunk, phrase) for phrase in phrases):
            relevant_chunks += 1
            if first_relevant_rank is None:
                first_relevant_rank = index

    phrase_chunk_rate = relevant_chunks / len(chunks) if chunks else 0.0
    any_phrase_hit = 1.0 if first_relevant_rank is not None else 0.0
    first_phrase_rank_reciprocal = (
        1.0 / first_relevant_rank if first_relevant_rank is not None else 0.0
    )

    return RagBenchmarkScores(
        phrase_coverage=round(phrase_coverage, 4),
        phrase_chunk_rate=round(phrase_chunk_rate, 4),
        any_phrase_hit=round(any_phrase_hit, 4),
        first_phrase_rank_reciprocal=round(first_phrase_rank_reciprocal, 4),
        expected_phrase_count=len(phrases),
        matched_phrase_count=len(matched),
        retrieved_chunk_count=len(chunks),
    )


def build_rag_evaluation_context(
    *,
    retrieval_mode: RetrievalMode,
    retrieve_limit: int,
    rerank_enabled: bool,
    rerank_top_n: int,
    chunks: list[ChunkHit],
    rerank_applied: bool,
    hybrid_fts_active: bool,
    chunk_size: int | None = None,
    chunk_overlap: int | None = None,
    indexed_chunk_count: int | None = None,
) -> RagEvaluationContext:
    """Build evaluation context describing the retrieval run configuration."""
    return RagEvaluationContext(
        retrieval_mode=retrieval_mode,
        retrieve_limit=retrieve_limit,
        rerank_enabled=rerank_enabled,
        rerank_top_n=rerank_top_n,
        effective_k=len(chunks),
        score_kind=resolve_score_kind(
            rerank_applied=rerank_applied,
            retrieval_mode=retrieval_mode,
            hybrid_fts_active=hybrid_fts_active,
        ),
        chunk_size=chunk_size,
        chunk_overlap=chunk_overlap,
        indexed_chunk_count=indexed_chunk_count,
        hybrid_fts_active=hybrid_fts_active,
        rerank_applied=rerank_applied,
    )


def compute_retrieval_proxy_metrics(
    *,
    chunks: list[ChunkHit],
    merged_context: str,
    score_kind: ScoreKind,
) -> RetrievalProxyMetrics:
    """Compute retrieval-only quality proxies without ground-truth phrases."""
    effective_k = len(chunks)
    if not chunks:
        return RetrievalProxyMetrics(
            chunk_count=0,
            mean_chunk_score=0.0,
            max_chunk_score=0.0,
            context_length_chars=len(merged_context),
            score_kind=score_kind,
            effective_k=effective_k,
        )
    scores = [chunk.score for chunk in chunks]
    return RetrievalProxyMetrics(
        chunk_count=len(chunks),
        mean_chunk_score=round(sum(scores) / len(scores), 4),
        max_chunk_score=round(max(scores), 4),
        context_length_chars=len(merged_context),
        score_kind=score_kind,
        effective_k=effective_k,
    )

File: ui/src/lib/ragBenchmarks.ts
import type { WorkflowTraceStep } from "../api/workflows";

export type ScoreKind = "cosine" | "rrf" | "reranker";

export type RagEvaluationContext = {
  retrieval_mode: "vector" | "hybrid";
  retrieve_limit: number;
  rerank_enabled: boolean;
  rerank_top_n: number;
  effective_k: number;
  score_kind: ScoreKind;
  chunk_size?: number | null;
  chunk_overlap?: number | null;
  indexed_chunk_count?: number | null;
  hybrid_fts_active?: boolean;
  rerank_applied?: boolean;
};

export type RagWorkflowRunMeta = {
  workflowId: string;
  ragBenchmarks?: Record<string, number>;
  ragEvaluationContext?: RagEvaluationContext | null;
  validationPassed?: boolean;
  validationErrors?: string[];
  matchedPhrases?: string[];
  missingPhrases?: string[];
  indexedChunkCount?: number;
  documentTitle?: string;
  documentSource?: string;
  expectedPhrases?: string[];
};

export type RagQualityMetrics = {
  phrase_coverage?: number;
  phrase_chunk_rate?: number;
  any_phrase_hit?: number;
  first_phrase_rank_reciprocal?: number;
  expected_phrase_count?: number;
  matched_phrase_count?: number;
  retrieved_chunk_count?: number;
  gold_semantic_relevance?: number;
  mean_gold_semantic_relevance?: number;
  gold_semantic_precision?: number;
  gold_semantic_rank_reciprocal?: number;
};

export type RetrievalProxyMetrics = {
  chunk_count?: number;
  mean_chunk_score?: number;
  max_chunk_score?: number;
  context_length_chars?: number;
  score_kind?: ScoreKind;
  effective_k?: number;
};

export type RagDashboardData = {
  mode: "validation" | "retrieval";
  quality: RagQualityMetrics;
  retrieval: RetrievalProxyMetrics;
  evaluationContext: RagEvaluationContext | null;
  validationPassed?: boolean;
  validationErrors?: string[];
  matchedPhrases?: string[];
  missingPhrases?: string[];
  indexedChunkCount?: number;
  documentTitle?: string;
  documentSource?: string;
  expectedPhrases?: string[];
};

export type MetricDefinition = {
  key: keyof RagQualityMetrics | keyof RetrievalProxyMetrics;
  label: string;
  description: string;
  format: "percent" | "score" | "count" | "chars";
  source: "quality" | "retrieval";
};

export const VALIDATION_METRICS: MetricDefinition[] = [
  {
    key: "gold_semantic_relevance",
    label: "Gold semantic relevance",
    description: "Max cosine similarity between the gold answer and any retrieved chunk embedding",
    format: "score",
    source: "quality",
  },
  {
    key: "gold_semantic_precision",
    label: "Gold semantic precision",
    description: "Share of retrieved chunks whose embeddings exceed the relevance threshold vs gold",
    format: "percent",
    source: "quality",
  },
  {
    key: "mean_gold_semantic_relevance",
    label: "Mean gold semantic relevance",
    description: "Average cosine similarity between the gold answer and retrieved chunk embeddings",
    format: "score",
    source: "quality",
  },
  {
    key: "gold_semantic_rank_reciprocal",
    label: "Semantic rank (1/r)",
    description: "Reciprocal rank of the first chunk semantically relevant to the gold answer",
    format: "percent",
    source: "quality",
  },
  {
    key: "phrase_coverage",
    label: "Phrase coverage",
    description: "Share of expected phrases found as substrings in retrieved chunks",
    format: "percent",
    source: "quality",
  },
  {
    key: "phrase_chunk_rate",
    label: "Phrase-bearing chunk rate",
    description: "Share of retrieved chunks that contain at least one expected phrase",
    format: "percent",
    source: "quality",
  },
  {
    key: "any_phrase_hit",
    label: "Any-phrase hit",
    description: "Binary: 1 when any chunk contains an expected phrase (k = effective chunk count)",
    format: "percent",
    source: "quality",
  },
  {
    key: "first_phrase_rank_reciprocal",
    label: "First-phrase rank (1/r)",
    description: "Reciprocal rank of the first chunk containing any expected phrase",
    format: "percent",
    source: "quality",
  },
];

export const RETRIEVAL_METRICS: MetricDefinition[] = [
  {
    key: "max_chunk_score",
    label: "Max chunk score",
    description: "Highest score among retrieved chunks (semantics depend on score kind)",
    format: "score",
    source: "retrieval",
  },
  {
    key: "mean_chunk_score",
    label: "Mean chunk score",
    description: "Average score across retrieved chunks (semantics depend on score kind)",
    format: "score",
    source: "retrieval",
  },
  {
    key: "chunk_count",
    label: "Chunks retrieved",
    description: "Number of chunks returned by retrieval",
    format: "count",
    source: "retrieval",
  },
  {
    key: "context_length_chars",
    label: "Context size",
    description: "Character length of merged context passed downstream",
    format: "chars",
    source: "retrieval",
  },
];

const SCORE_KIND_LABELS: Record<ScoreKind, string> = {
  cosine: "cosine similarity",
  rrf: "hybrid RRF",
  reranker: "cross-encoder reranker",
};

export function scoreKindLabel(kind: ScoreKind | undefined): string {
  if (!kind) {
    return "unknown";
  }
  return SCORE_KIND_LABELS[kind];
}

function asRecord(value: unknown): Record<string, number> | null {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    return null;
  }
  const record: Record<string, number> = {};
  for (const [key, item] of Object.entries(value)) {
    if (typeof item === "number" && Number.isFinite(item)) {
      record[key] = item;
    }
  }
  return Object.keys(record).length > 0 ? record : null;
}

function asEvaluationContext(value: unknown): RagEvaluationContext | null {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    return null;
  }
  const raw = value as Record<string, unknown>;
  const scoreKind = raw.score_kind;
  if (scoreKind !== "cosine" && scoreKind !== "rrf" && scoreKind !== "reranker") {
    return null;
  }
  const retrievalMode = raw.retrieval_mode;
  if (retrievalMode !== "vector" && retrievalMode !== "hybrid") {
    return null;
  }
  return {
    retrieval_mode: retrievalMode,
    retrieve_limit: Number(raw.retrieve_limit),
    rerank_enabled: Boolean(raw.rerank_enabled),
    rerank_top_n: Number(raw.rerank_top_n),
    effective_k: Number(raw.effective_k),
    score_kind: scoreKind,
    chunk_size: typeof raw.chunk_size === "number" ? raw.chunk_size : null,
    chunk_overlap: typeof raw.chunk_overlap === "number" ? raw.chunk_overlap : null,
    indexed_chunk_count:
      typeof raw.indexed_chunk_count === "number" ? raw.indexed_chunk_count : null,
  };
}

function pickMetrics<T extends Record<string, number>>(
  source: Record<string, number> | null | undefined,
  keys: (keyof T)[],
): Partial<T> {
  if (!source) {
    return {};
  }
  const picked: Partial<T> = {};
  for (const key of keys) {
    const value = source[String(key)];
    if (typeof value === "number") {
      picked[key] = value as T[keyof T];
    }
  }
  return picked;
}

export function extractRagBenchmarksFromTrace(trace: WorkflowTraceStep[]): RagQualityMetrics {
  const validateStep = [...trace]
    .reverse()
    .find((step) => step.node_id === "validate_retrieval");
  const fromTrace = asRecord(validateStep?.output_update?.rag_benchmarks);
  return pickMetrics<RagQualityMetrics>(fromTrace, [
    "gold_semantic_relevance",
    "mean_gold_semantic_relevance",
    "gold_semantic_precision",
    "gold_semantic_rank_reciprocal",
    "phrase_coverage",
    "phrase_chunk_rate",
    "any_phrase_hit",
    "first_phrase_rank_reciprocal",
    "expected_phrase_count",
    "matched_phrase_count",
    "retrieved_chunk_count",
  ]);
}

export function extractRetrievalMetricsFromTrace(trace: WorkflowTraceStep[]): RetrievalProxyMetrics {
  const mergeStep = [...trace].reverse().find((step) => step.node_id === "merge_context");
  const fromTrace = asRecord(mergeStep?.output_update?.retrieval_metrics);
  const metrics = pickMetrics<{
    chunk_count?: number;
    mean_chunk_score?: number;
    max_chunk_score?: number;
    context_length_chars?: number;
    effective_k?: number;
  }>(fromTrace, [
    "chunk_count",
    "mean_chunk_score",
    "max_chunk_score",
    "context_length_chars",
    "effective_k",
  ]);
  const scoreKind = mergeStep?.output_update?.retrieval_metrics;
  const rawKind =
    scoreKind && typeof scoreKind === "object" && !Array.isArray(scoreKind)
      ? (scoreKind as Record<string, unknown>).score_kind
      : undefined;
  if (rawKind === "cosine" || rawKind === "rrf" || rawKind === "reranker") {
    return { ...metrics, score_kind: rawKind };
  }
  return metrics;
}

export function extractEvaluationContextFromTrace(
  trace: WorkflowTraceStep[],
): RagEvaluationContext | null {
  const mergeStep = [...trace].reverse().find((step) => step.node_id === "merge_context");
  const fromMerge = asEvaluationContext(mergeStep?.output_update?.rag_evaluation_context);
  if (fromMerge) {
    return fromMerge;
  }
  const validateStep = [...trace]
    .reverse()
    .find((step) => step.node_id === "validate_retrieval");
  return asEvaluationContext(validateStep?.output_update?.rag_evaluation_context);
}

export function hasQualityMetrics(metrics: RagQualityMetrics): boolean {
  return (
    metrics.gold_semantic_relevance !== undefined ||
    metrics.gold_semantic_precision !== undefined ||
    metrics.mean_gold_semantic_relevance !== undefined ||
    metrics.phrase_coverage !== undefined ||
    metrics.phrase_chunk_rate !== undefined ||
    metrics.any_phrase_hit !== undefined ||
    metrics.first_phrase_rank_reciprocal !== undefined
  );
}

export function hasRetrievalMetrics(metrics: RetrievalProxyMetrics): boolean {
  return (
    metrics.chunk_count !== undefined ||
    metrics.mean_chunk_score !== undefined ||
    metrics.max_chunk_score !== undefined ||
    metrics.context_length_chars !== undefined
  );
}

const DEFAULT_SCORE_THRESHOLDS = { good: 0.75, warn: 0.45 };
const RRF_SCORE_THRESHOLDS = { good: 0.02, warn: 0.01 };
const RERANK_SCORE_THRESHOLDS = { good: 0.5, warn: 0.02 };

export function scoreThresholdsForKind(scoreKind: ScoreKind | undefined): {
  good: number;
  warn: number;
} {
  if (scoreKind === "rrf") {
    return RRF_SCORE_THRESHOLDS;
  }
  if (scoreKind === "reranker") {
    return RERANK_SCORE_THRESHOLDS;
  }
  return DEFAULT_SCORE_THRESHOLDS;
}

export function scoreBand(
  value: number,
  format: MetricDefinition["format"],
  scoreKind?: ScoreKind,
): "good" | "warn" | "bad" {
  if (format === "count" || format === "chars") {
    return value > 0 ? "good" : "bad";
  }
  if (format === "score") {
    const thresholds = scoreThresholdsForKind(scoreKind);
    if (value >= thresholds.good) {
      return "good";
    }
    if (value >= thresholds.warn) {
      return "warn";
    }
    return "bad";
  }
  if (value >= 0.8) {
    return "good";
  }
  if (value >= 0.5) {
    return "warn";
  }
  return "bad";
}

export function formatMetricValue(value: number | undefined, format: MetricDefinition["format"]): string {
  if (value === undefined) {
    return "—";
  }
  if (format === "percent") {
    return `${Math.round(value * 100)}%`;
  }
  if (format === "score") {
    return value.toFixed(3);
  }
  if (format === "chars") {
    if (value >= 1000) {
      return `${(value / 1000).toFixed(1)}k chars`;
    }
    return `${Math.round(value)} chars`;
  }
  return String(Math.round(value));
}

export function primaryValidationScore(metrics: RagQualityMetrics): number | undefined {
  if (metrics.gold_semantic_relevance !== undefined) {
    return metrics.gold_semantic_relevance;
  }
  if (metrics.mean_gold_semantic_relevance !== undefined) {
    return metrics.mean_gold_semantic_relevance;
  }
  return (
    metrics.phrase_coverage ??
    metrics.phrase_chunk_rate ??
    metrics.any_phrase_hit ??
    metrics.first_phrase_rank_reciprocal
  );
}

export function buildRagDashboardData(input: {
  workflowId: string;
  trace: WorkflowTraceStep[];
  ragBenchmarks?: Record<string, number>;
  ragEvaluationContext?: RagEvaluationContext | null;
  validationPassed?: boolean;
  validationErrors?: string[];
  matchedPhrases?: string[];
  missingPhrases?: string[];
  indexedChunkCount?: number;
  documentTitle?: string;
  documentSource?: string;
  expectedPhrases?: string[];
}): RagDashboardData | null {
  const qualityFromApi = pickMetrics<RagQualityMetrics>(input.ragBenchmarks, [
    "gold_semantic_relevance",
    "mean_gold_semantic_relevance",
    "gold_semantic_precision",
    "gold_semantic_rank_reciprocal",
    "phrase_coverage",
    "phrase_chunk_rate",
    "any_phrase_hit",
    "first_phrase_rank_reciprocal",
    "expected_phrase_count",
    "matched_phrase_count",
    "retrieved_chunk_count",
  ]);
  const qualityFromTrace = extractRagBenchmarksFromTrace(input.trace);
  const quality: RagQualityMetrics = { ...qualityFromTrace, ...qualityFromApi };
  const retrieval = extractRetrievalMetricsFromTrace(input.trace);
  const evaluationContext =
    input.ragEvaluationContext ?? extractEvaluationContextFromTrace(input.trace);

  const isValidation = input.workflowId === "rag-validation";
  const isRetrieval = input.workflowId === "rag-retrieval";

  if (!isValidation && !isRetrieval) {
    return null;
  }
  if (!hasQualityMetrics(quality) && !hasRetrievalMetrics(retrieval)) {
    return null;
  }

  return {
    mode: isValidation ? "validation" : "retrieval",
    quality,
    retrieval,
    evaluationContext,
    validationPassed: input.validationPassed,
    validationErrors: input.validationErrors,
    matchedPhrases: input.matchedPhrases,
    missingPhrases: input.missingPhrases,
    indexedChunkCount: input.indexedChunkCount,
    documentTitle: input.documentTitle,
    documentSource: input.documentSource,
    expectedPhrases: input.expectedPhrases,
  };
}
